{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1c1bdcb5",
   "metadata": {},
   "source": [
    "# SeqApiPop analyses: RFMix\n",
    "\n",
    "- Phasing with SNPs\n",
    "    - Select SNPs\n",
    "    - VCFs per Chromosme\n",
    "    - Phasing with Shapeit\n",
    "        - Phasing\n",
    "        - Convert phased genotype back to vcf\n",
    "        - Concatenate VCFs\n",
    "        \n",
    "- RFMix\n",
    "- Select reference and query Samples\n",
    "    - Make vcf files\n",
    "- Make genetic maps\n",
    "    - Find crossing-over in data from Liu et al., 2015\n",
    "- Run RFMix\n",
    "    - Couldn't get chromosome 1 to work\n",
    "    \n",
    "    \n",
    "## Phasing with shapeit\n",
    "\n",
    "### Select SNPs\n",
    "\n",
    "- MAF > 1\n",
    "- Max alleles = 2\n",
    "- Chromosomes as numbers\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b0cd93c",
   "metadata": {},
   "source": [
    "Firstly, the names of the chromosomes must be changed to numbers from 1 to 16 because we do not include the mitochondrial chromosome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ecf26d0",
   "metadata": {},
   "source": [
    "here is the script \"ChangeChr\" : to change the chromosome names into numbers:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa35079e",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/sh\n",
    "\n",
    "\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "\n",
    "\n",
    "VCF=MetaGenotypesCalled403_raw_snps_filtre_isec_plink.vcf.gz\n",
    "\n",
    "\n",
    "echo \"NC_037638.1 1\" >> chr_name_conv.txt\n",
    "echo \"NC_037639.1 2\" >> chr_name_conv.txt\n",
    "echo \"NC_037640.1 3\" >> chr_name_conv.txt\n",
    "echo \"NC_037641.1 4\" >> chr_name_conv.txt\n",
    "echo \"NC_037642.1 5\" >> chr_name_conv.txt\n",
    "echo \"NC_037643.1 6\" >> chr_name_conv.txt\n",
    "echo \"NC_037644.1 7\" >> chr_name_conv.txt\n",
    "echo \"NC_037645.1 8\" >> chr_name_conv.txt\n",
    "echo \"NC_037646.1 9\" >> chr_name_conv.txt\n",
    "echo \"NC_037647.1 10\" >> chr_name_conv.txt\n",
    "echo \"NC_037648.1 11\" >> chr_name_conv.txt\n",
    "echo \"NC_037649.1 12\" >> chr_name_conv.txt\n",
    "echo \"NC_037650.1 13\" >> chr_name_conv.txt\n",
    "echo \"NC_037651.1 14\" >> chr_name_conv.txt\n",
    "echo \"NC_037652.1 15\" >> chr_name_conv.txt\n",
    "echo \"NC_037653.1 16\" >> chr_name_conv.txt\n",
    "echo \"NC_001566.1 17\" >> chr_name_conv.txt\n",
    "\n",
    "\n",
    "bcftools annotate --rename-chrs chr_name_conv.txt $VCF | bgzip > rename.vcf.gz\n",
    "\n",
    "bcftools index rename.vcf.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef504502",
   "metadata": {},
   "source": [
    "Here, the script for the selection of SNPs:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85d7aeeb",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#makeVCF.bash\n",
    "\n",
    "#select from the vcf with chromosomes indicated as numbers, for plinkAnalyses\n",
    "#filter on MAF > 0.01\n",
    "#nb alleles max = 2\n",
    "\n",
    "module load -f /home/gencel/vignal/save/000_ProgramModules/program_module\n",
    "\n",
    "#Select the 403 samples and filter on MAF001\n",
    "\n",
    "bcftools view  --samples-file /home/agirardon/work/seqapipopOnHAV3_1/RFmix/IndsAll403.list \\\n",
    "            --min-af 0.01:minor \\\n",
    "            --max-alleles 2 \\\n",
    "            --output-file /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink.vcf.gz \\\n",
    "            --output-type z \\\n",
    "            /home/agirardon/work/seqapipopOnHAV3_1/RFmix/rename.vcf.gz\n",
    "bcftools index /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink.vcf.gz \n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1197725",
   "metadata": {},
   "source": [
    "### VCFs per chromosome\n",
    "\n",
    "- Shapeit requires one vcf per chromosome\n",
    "\n",
    "Indeed, Shapeit needs an ascending order for SNPs positions, and in the VCF with all chromosomes the positions are in ascending order for one chromosome, but when moving to the next chromosome the position returns to the beginning of that chromosome "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cefb5ebe",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#separateChromosomes.sh\n",
    "\n",
    "module load -f /home/gencel/vignal/save/000_ProgramModules/program_module\n",
    "\n",
    "\n",
    "#Select the 403 samples and filter on MAF001\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "bcftools view  --regions ${i} \\\n",
    "        --output-file /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf.gz \\\n",
    "       /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink.vcf.gz\n",
    "bcftools index -f /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf.gz\n",
    "done\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "do\n",
    "bcftools index -f /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf.gz\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cbab417",
   "metadata": {},
   "source": [
    "### Phasing with Shapeit\n",
    "\n",
    "#### Phasing\n",
    "\n",
    "Remark: the vcf file will be renamed, because Shapeit unzip the files, if they have .gz in their name. that's why .gz is only removed from the VCF name."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f75c3900",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#phasingShapeit.bash\n",
    "\n",
    "module load bioinfo/shapeit.v2.904\n",
    "\n",
    "\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "sbatch --mem=50g --wrap=\"shapeit --input-vcf SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf --force -O SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf\"\n",
    "\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43da1585",
   "metadata": {},
   "source": [
    "#### Convert phased genotypes back to vcf\n",
    "\n",
    "Each loop was launched one after the other \n",
    "\n",
    "``` bash\n",
    "#!/bin/bash\n",
    "\n",
    "#convertToVcf.bash\n",
    "\n",
    "module load bioinfo/shapeit.v2.904\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "\n",
    "\n",
    "sbatch --wrap=\"shapeit -convert --input-haps SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf --output-vcf SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf\"\n",
    "\n",
    "done\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "sbatch --wrap=\"bgzip SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba5c75f8",
   "metadata": {},
   "source": [
    "#### Concatenate VCFs \n",
    "\n",
    "\n",
    "- Make list\n",
    "\n",
    "```bash\n",
    "#!/bin/sh \n",
    "\n",
    "ls ~/work/seqapipopOnHAV3_1/RFmix/out/*phased*.vcf.gz > vcf.list\n",
    "\n",
    "```\n",
    "\n",
    "File vcf.list look like : \n",
    "\n",
    "```\n",
    "/home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_chr10.vcf.gz\n",
    "/home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_chr11.vcf.gz\n",
    "/home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_chr12.vcf.gz\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f50fbc9",
   "metadata": {},
   "source": [
    "The chromosomes do not come out in the right order, just cut and paste to reorder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "870ecbff",
   "metadata": {},
   "source": [
    "## RFMix \n",
    "\n",
    "### Select reference and query Samples\n",
    "\n",
    "Select from an Admixture Q matrix with K = 3 the individuals with > 0.93 pure backgrounds as reference => IndsReference.list The other samples => IndsQuery.list\n",
    "\n",
    "#### Make BCF files\n",
    "\n",
    "- References\n",
    "\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#selectBcfRef.bash\n",
    "\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "#Select the 403 samples and filter on MAF001\n",
    "\n",
    "bcftools view  --samples-file ~/work/seqapipopOnHAV3_1/RFmix/IndsAll403ref.list  \\\n",
    "        --output-file ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_320_MAF001_diAllelic_phased_Ref.bcf \\\n",
    "        --output-type b \\\n",
    "        ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_All.vcf.gz\n",
    "bcftools index SeqApiPop_320_MAF001_diAllelic_phased_Ref.bcf\n",
    "```\n",
    "\n",
    "- Query\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#selectBcfQuery.bash\n",
    "\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "bcftools view  --samples-file ~/work/seqapipopOnHAV3_1/RFmix/IndsAll403_query.list \\\n",
    "            --output-file ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_Query.bcf \\\n",
    "            --output-type b \\\n",
    "            ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_629_MAF001_diAllelic_phased_All.vcf.gz\n",
    "bcftools index SeqApiPop_403_MAF001_diAllelic_phased_Query.bcf\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63d06910",
   "metadata": {},
   "source": [
    "## Make a genetic maps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2f99d11",
   "metadata": {},
   "source": [
    "To make the genetic map, here we simply recovered the same genetic map that Alain Vignal used for his RFmix analysis \n",
    "\n",
    "Find crossing-overs in data from Liu et al., 2015\n",
    "\n",
    "Liu H, Zhang X, Huang J, Chen J-Q, Tian D, Hurst LD, et al. Causes and consequences of crossing-over evidenced via a high-resolution recombinational landscape of the honey bee. Genome Biol. 2015 Jan 1;16:15.\n",
    "\n",
    "Reads from the project SRP043350 (Liu et al., 2015) were retrieved from Short Read Archive (SRA) (https://www.ncbi.nlm.nih.gov/sra), aligned to the reference genome for SNP detection (GATK pipeline): vcf file LiuGenotypesSNPs2allelesForCoSearchDuplOK.vcf.gz\n",
    "\n",
    "The script find_crossing_overs.py was then used to detect crossing overs:\n",
    "\n",
    "    -l Colony*.list : list of samples in the Colony\n",
    "    -q SRR* : the colony's queen\n",
    "    -c 6 : number of crossing-overs allowed simultaneously over all individuals\n",
    "\n",
    "Then a genetic map was constructed from the data:\n",
    "\n",
    "After importing the 3 files Colony3.list, make table with one line when multiple COs at one position and add column indicating the lines that are involved in double COs.\n",
    "\n",
    "   - A double CO is due either to:\n",
    "        - A non-crossing-over event marked by a single SNP\n",
    "        - non-crossing-over events marked by more than one SNP will be detected by the distance to the previous and/or the next CO in the same individual. The distance is set here at 10 kb, as suggested by Liu, et al., 2015.\n",
    "        -  A genotyping error (especially in the case of nb_COs = 1 or 2)\n",
    "        -  nb_COs = number of ofspring from a colony, with COs at the same position\n",
    "        - Problems in the assembly, such as small inversions, especially in the case of nb_COs > 2)\n",
    "   - Detection of double crossing-overs: two lines must be removed\n",
    "       - Line for the recombinant before the SNP : the previous and the next vectors are identical\n",
    "       - Line for the recombinant after the SNP : the current examined vector and the vectors two steps back are identical\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "051238a5",
   "metadata": {},
   "source": [
    "## Run RFmix\n",
    "\n",
    "Add population colums to IndsAll403ref.list -> IndsAll403popref.list\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c0338b5",
   "metadata": {},
   "source": [
    "```\n",
    "head IndsAll403popref\n",
    "\n",
    "Ab-PacBio\tBlack\n",
    "BER10\tYellow\n",
    "BER11\tYellow\n",
    "BER12\tYellow\n",
    "BER13\tYellow\n",
    "BER14\tYellow\n",
    "BER15\tYellow\n",
    "BER16\tYellow\n",
    "BER18\tYellow\n",
    "BER19\tYellow\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7457b35",
   "metadata": {},
   "source": [
    "So we create a script to run the RFmix script : LanceRunRFmix.sh\n",
    "    \n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#LanceRunRFMix.bash\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "do\n",
    "sbatch --mem=30g ~/work/seqapipopOnHAV3_1/RFmix/runRFMixWithGenetMap.bash ${i}\n",
    "\n",
    "done\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4da8904",
   "metadata": {},
   "source": [
    "This allows us to run the following script, launching the RFmix on all our chromosomes : runRFMixWithGenetMap.bash\n",
    "\n",
    "\n",
    "\n",
    "```bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "module load bioinfo/rfmix-9505bfa\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "rfmix -f ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_Query.bcf \\\n",
    "        -r ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_320_MAF001_diAllelic_phased_Ref.bcf \\\n",
    "        --chromosome=${1} \\\n",
    "        -m ~/work/seqapipopOnHAV3_1/RFmix/IndsAll403popref.list \\\n",
    "        -g ~/work/seqapipopOnHAV3_1/RFmix/GenetMap_march_2021_AV.txt \\\n",
    "        -o ~/work/seqapipopOnHAV3_1/RFmix/out2/SeqApiPopRFmixChr${1}\n",
    "        --reanalyze-reference \n",
    "    \n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec9b5ed9",
   "metadata": {},
   "source": [
    "Example output :\n",
    "    \n",
    "- SeqApiPopRFmixChr10.fb.tsv\n",
    "- SeqApiPopRFmixChr10.msp.tsv\n",
    "- SeqApiPopRFmixChr10.rfmix.Q\n",
    "- SeqApiPopRFmixChr10.sis.tsv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
