{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f79f0ba0",
   "metadata": {},
   "source": [
    "# Analyses SeqApiPop : RFMix\n",
    "\n",
    "- Mise en phase avec les SNP\n",
    "    - Sélectionner les SNP\n",
    "    - VCFs par Chromosme\n",
    "    - Phaser avec Shapeit\n",
    "        - Phasage\n",
    "        - Convertir le génotype phasé en vcf\n",
    "        - Concaténation des VCFs\n",
    "        \n",
    "- RFMix\n",
    "- Sélectionner les échantillons de référence et d'interrogation\n",
    "    - Créer des fichiers vcf\n",
    "- Créer des cartes génétiques\n",
    "    - Trouver le crossing-over dans les données de Liu et al. 2015\n",
    "- Exécuter RFMix\n",
    "    - Impossible de faire fonctionner le chromosome 1.\n",
    "    \n",
    "    \n",
    "### Mise en phase avec shapeit\n",
    "\n",
    "### Sélectionner les SNP\n",
    "\n",
    "- MAF > 1\n",
    "- Max allèles = 2\n",
    "- Chromosomes en tant que nombres\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1de4fffd",
   "metadata": {},
   "source": [
    "Tout d'abord, les noms des chromosomes doivent être changés en nombres de 1 à 16 car nous n'incluons pas le chromosome mitochondrial\n",
    "\n",
    "voici le script \"ChangeChr\" : pour changer les noms des chromosomes en numéros :\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "984fd7a5",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/sh\n",
    "\n",
    "\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "\n",
    "\n",
    "VCF=MetaGenotypesCalled403_raw_snps_filtre_isec_plink.vcf.gz\n",
    "\n",
    "\n",
    "echo \"NC_037638.1 1\" >> chr_name_conv.txt\n",
    "echo \"NC_037639.1 2\" >> chr_name_conv.txt\n",
    "echo \"NC_037640.1 3\" >> chr_name_conv.txt\n",
    "echo \"NC_037641.1 4\" >> chr_name_conv.txt\n",
    "echo \"NC_037642.1 5\" >> chr_name_conv.txt\n",
    "echo \"NC_037643.1 6\" >> chr_name_conv.txt\n",
    "echo \"NC_037644.1 7\" >> chr_name_conv.txt\n",
    "echo \"NC_037645.1 8\" >> chr_name_conv.txt\n",
    "echo \"NC_037646.1 9\" >> chr_name_conv.txt\n",
    "echo \"NC_037647.1 10\" >> chr_name_conv.txt\n",
    "echo \"NC_037648.1 11\" >> chr_name_conv.txt\n",
    "echo \"NC_037649.1 12\" >> chr_name_conv.txt\n",
    "echo \"NC_037650.1 13\" >> chr_name_conv.txt\n",
    "echo \"NC_037651.1 14\" >> chr_name_conv.txt\n",
    "echo \"NC_037652.1 15\" >> chr_name_conv.txt\n",
    "echo \"NC_037653.1 16\" >> chr_name_conv.txt\n",
    "echo \"NC_001566.1 17\" >> chr_name_conv.txt\n",
    "\n",
    "\n",
    "bcftools annotate --rename-chrs chr_name_conv.txt $VCF | bgzip > rename.vcf.gz\n",
    "\n",
    "bcftools index rename.vcf.gz\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "470c9836",
   "metadata": {},
   "source": [
    "Ici, le script pour la sélection des SNP :"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7ad4cc9",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#makeVCF.bash\n",
    "\n",
    "#select from the vcf with chromosomes indicated as numbers, for plinkAnalyses\n",
    "#filter on MAF > 0.01\n",
    "#nb alleles max = 2\n",
    "\n",
    "module load -f /home/gencel/vignal/save/000_ProgramModules/program_module\n",
    "\n",
    "#Select the 403 samples and filter on MAF001\n",
    "\n",
    "bcftools view  --samples-file /home/agirardon/work/seqapipopOnHAV3_1/RFmix/IndsAll403.list \\\n",
    "            --min-af 0.01:minor \\\n",
    "            --max-alleles 2 \\\n",
    "            --output-file /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink.vcf.gz \\\n",
    "            --output-type z \\\n",
    "            /home/agirardon/work/seqapipopOnHAV3_1/RFmix/rename.vcf.gz\n",
    "bcftools index /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink.vcf.gz \n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c4c56d2",
   "metadata": {},
   "source": [
    "### VCFs par chromosome\n",
    "\n",
    "- Shapeit requiert un VCF par chromosome.\n",
    "\n",
    "En effet, Shapeit a besoin d'un ordre ascendant pour les positions des SNPs, et dans le VCF avec tous les chromosomes les positions sont dans l'ordre ascendant pour un chromosome, mais quand on passe au chromosome suivant la position revient au début de ce chromosome "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0b49a48",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#separateChromosomes.sh\n",
    "\n",
    "module load -f /home/gencel/vignal/save/000_ProgramModules/program_module\n",
    "\n",
    "\n",
    "#Select the 403 samples and filter on MAF001\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "bcftools view  --regions ${i} \\\n",
    "        --output-file /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf.gz \\\n",
    "       /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink.vcf.gz\n",
    "bcftools index -f /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf.gz\n",
    "done\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "do\n",
    "bcftools index -f /home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf.gz\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52448636",
   "metadata": {},
   "source": [
    "### Phasage avec Shapeit\n",
    "\n",
    "#### Mise en phase\n",
    "\n",
    "Remarque : le fichier vcf sera renommé, parce que Shapeit décompresse les fichiers, s'ils ont .gz dans leur nom. C'est pourquoi .gz est seulement enlevé du nom VCF."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3a5d4a2",
   "metadata": {},
   "source": [
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#phasingShapeit.bash\n",
    "\n",
    "module load bioinfo/shapeit.v2.904\n",
    "\n",
    "\n",
    "\n",
    "for i in  $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "sbatch --mem=50g --wrap=\"shapeit --input-vcf SeqApipop_403_MAF001_diAllelic_plink_chr${i}.vcf --force -O SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf\"\n",
    "\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76dacc12",
   "metadata": {},
   "source": [
    "#### Convertir les génotypes en phase en vcf\n",
    "\n",
    "Chaque boucle a été lancée l'une après l'autre \n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#convertToVcf.bash\n",
    "\n",
    "module load bioinfo/shapeit.v2.904\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "for i in $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "\n",
    "\n",
    "sbatch --wrap=\"shapeit -convert --input-haps SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf --output-vcf SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf\"\n",
    "\n",
    "done\n",
    "\n",
    "for i in $(seq 1 16)\n",
    "\n",
    "do\n",
    "\n",
    "sbatch --wrap=\"bgzip SeqApiPop_403_MAF001_diAllelic_phased_chr${i}.vcf\"\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f5bceeb",
   "metadata": {},
   "source": [
    "#### Concaténation de VCFs \n",
    "\n",
    "\n",
    "- Faire une liste\n",
    "\n",
    "```bash\n",
    "#!/bin/sh \n",
    "\n",
    "ls ~/work/seqapipopOnHAV3_1/RFmix/out/*phased*.vcf.gz > vcf.list\n",
    "\n",
    "```\n",
    "\n",
    "Le fichier vcf.list ressemble à : \n",
    "\n",
    "```\n",
    "/home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_chr10.vcf.gz\n",
    "/home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_chr11.vcf.gz\n",
    "/home/agirardon/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_chr12.vcf.gz\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bca8d308",
   "metadata": {},
   "source": [
    "Les chromosomes ne sortent pas dans le bon ordre, il suffit de couper et coller pour les réordonner."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a141e4e",
   "metadata": {},
   "source": [
    "## RFMix \n",
    "\n",
    "### Select reference and query Samples\n",
    "\n",
    "Select from an Admixture Q matrix with K = 3 the individuals with > 0.93 pure backgrounds as reference => IndsReference.list The other samples => IndsQuery.list\n",
    "\n",
    "#### Make BCF files\n",
    "\n",
    "- References\n",
    "\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#selectBcfRef.bash\n",
    "\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "#Select the 403 samples and filter on MAF001\n",
    "\n",
    "bcftools view  --samples-file ~/work/seqapipopOnHAV3_1/RFmix/IndsAll403ref.list  \\\n",
    "        --output-file ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_320_MAF001_diAllelic_phased_Ref.bcf \\\n",
    "        --output-type b \\\n",
    "        ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_All.vcf.gz\n",
    "bcftools index SeqApiPop_320_MAF001_diAllelic_phased_Ref.bcf\n",
    "```\n",
    "\n",
    "- Query\n",
    "\n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#selectBcfQuery.bash\n",
    "\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "bcftools view  --samples-file ~/work/seqapipopOnHAV3_1/RFmix/IndsAll403_query.list \\\n",
    "            --output-file ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_Query.bcf \\\n",
    "            --output-type b \\\n",
    "            ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_629_MAF001_diAllelic_phased_All.vcf.gz\n",
    "bcftools index SeqApiPop_403_MAF001_diAllelic_phased_Query.bcf\n",
    "```\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebd4a5d7",
   "metadata": {},
   "source": [
    "## Faire une carte génétique"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b604be28",
   "metadata": {},
   "source": [
    "Pour faire la carte génétique, ici nous avons simplement récupéré la même carte génétique qu'Alain Vignal a utilisé pour son analyse RFmix. \n",
    "\n",
    "Trouver les crossing-overs dans les données de Liu et al. 2015\n",
    "\n",
    "Liu H, Zhang X, Huang J, Chen J-Q, Tian D, Hurst LD, et al. Causes et conséquences du crossing-over mis en évidence via un paysage recombinatoire à haute résolution de l'abeille domestique. Genome Biol. 2015 Jan 1;16:15.\n",
    "\n",
    "Les lectures du projet SRP043350 (Liu et al., 2015) ont été extraites de Short Read Archive (SRA) (https://www.ncbi.nlm.nih.gov/sra), alignées sur le génome de référence pour la détection des SNP (pipeline GATK) : fichier vcf LiuGenotypesNPs2allelesForCoSearchDuplOK.vcf.gz.\n",
    "\n",
    "Le script find_crossing_overs.py a ensuite été utilisé pour détecter les crossing overs :\n",
    "\n",
    "    -l Colonie*.list : liste des échantillons de la colonie\n",
    "    -q SRR* : la reine de la colonie\n",
    "    -c 6 : nombre de crossing-overs autorisés simultanément sur tous les individus.\n",
    "\n",
    "Ensuite, une carte génétique a été construite à partir de ces données :\n",
    "\n",
    "Après avoir importé les 3 fichiers Colony3.list, faire un tableau avec une ligne lorsque plusieurs COs à une position et ajouter une colonne indiquant les lignes qui sont impliquées dans les doubles COs.\n",
    "\n",
    "   - Un double CO est dû soit à :\n",
    "        - un événement de non-crossing-over marqué par un seul SNP\n",
    "        - Les événements de non-crossing-over marqués par plus d'un SNP seront détectés par la distance au CO précédent et/ou au CO suivant chez le même individu. La distance est fixée ici à 10 kb, comme suggéré par Liu, et al., 2015.\n",
    "        - Une erreur de génotypage (notamment dans le cas où nb_COs = 1 ou 2)\n",
    "        - nb_COs = nombre de descendants d'une colonie, avec des COs à la même position.\n",
    "        - Problèmes dans l'assemblage, comme de petites inversions, surtout dans le cas de nb_COs > 2)\n",
    "   - Détection des doubles crossing-overs : deux lignes doivent être retirées\n",
    "       - Ligne pour le recombinant avant le SNP : le vecteur précédent et le suivant sont identiques.\n",
    "       - Ligne pour le recombinant après le SNP : le vecteur actuel examiné et les vecteurs deux pas en arrière sont identiques.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25805e5e",
   "metadata": {},
   "source": [
    "## Exécuter RFmix\n",
    "\n",
    "Ajouter les colonnes de population à IndsAll403ref.list -> IndsAll403popref.list\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b8abb3",
   "metadata": {},
   "source": [
    "```\n",
    "head IndsAll403popref\n",
    "\n",
    "Ab-PacBio\tBlack\n",
    "BER10\tYellow\n",
    "BER11\tYellow\n",
    "BER12\tYellow\n",
    "BER13\tYellow\n",
    "BER14\tYellow\n",
    "BER15\tYellow\n",
    "BER16\tYellow\n",
    "BER18\tYellow\n",
    "BER19\tYellow\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "422f6455",
   "metadata": {},
   "source": [
    "Nous créons donc un script pour exécuter le script RFmix : LanceRunRFmix.sh\n",
    "    \n",
    "```bash\n",
    "#!/bin/bash\n",
    "\n",
    "#LanceRunRFMix.bash\n",
    "\n",
    "for i in $(seq 1 16)\n",
    "do\n",
    "sbatch --mem=30g ~/work/seqapipopOnHAV3_1/RFmix/runRFMixWithGenetMap.bash ${i}\n",
    "\n",
    "done\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72efa1c2",
   "metadata": {},
   "source": [
    "This allows us to run the following script, launching the RFmix on all our chromosomes : runRFMixWithGenetMap.bash\n",
    "\n",
    "\n",
    "\n",
    "```bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "module load bioinfo/rfmix-9505bfa\n",
    "module load bioinfo/bcftools-1.9\n",
    "\n",
    "rfmix -f ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_403_MAF001_diAllelic_phased_Query.bcf \\\n",
    "        -r ~/work/seqapipopOnHAV3_1/RFmix/out/SeqApiPop_320_MAF001_diAllelic_phased_Ref.bcf \\\n",
    "        --chromosome=${1} \\\n",
    "        -m ~/work/seqapipopOnHAV3_1/RFmix/IndsAll403popref.list \\\n",
    "        -g ~/work/seqapipopOnHAV3_1/RFmix/GenetMap_march_2021_AV.txt \\\n",
    "        -o ~/work/seqapipopOnHAV3_1/RFmix/out2/SeqApiPopRFmixChr${1}\n",
    "        --reanalyze-reference \n",
    "    \n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3079e996",
   "metadata": {},
   "source": [
    "Exemple d'output :\n",
    "    \n",
    "- SeqApiPopRFmixChr10.fb.tsv\n",
    "- SeqApiPopRFmixChr10.msp.tsv\n",
    "- SeqApiPopRFmixChr10.rfmix.Q\n",
    "- SeqApiPopRFmixChr10.sis.tsv"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
